_posts/2020-08-24-training-imagenet-with-r/training-imagenet-with-r.Rmd

```diff
@@ -0,0 +1,270 @@
+
+Now that we have resources capable of working with ImageNet, we need to find a place to download ImageNet from. Long story short, the easiest way is to use a variation of ImageNet used in the [ImageNet Large Scale Visual Recognition Challenge (ILSVRC)](http://www.image-net.org/challenges/LSVRC/), which contains a subset of about 250GB of data and can be easily downloaded from many [Kaggle](https://kaggle.com) competitions, like the [ImageNet Object Localization Challenge](https://www.kaggle.com/c/imagenet-object-localization-challenge).
+
+If you've read some of our previous posts, you might be thinking of using -- the [pins](https://pins.rstudio.com) package. The pins package enables you to cache, discover and share resources from many services, including Kaggle. You can learn more about data retrieval from Kaggle in the [Using Kaggle Boards](http://pins.rstudio.com/articles/boards-kaggle.html) article; in the meantime, let's assume you are already familiar with this package.
+
+All we need to do now is register the Kaggle board, retrieve ImageNet as a pin, and uncompress this file. Warning, the following code requires you to stare at a progress bar for, potentially, over an hour.
+
+```{r}
+library(pins)
+board_register("kaggle", token = "kaggle.json")
+
+pin_get("c/imagenet-object-localization-challenge", board = "kaggle")[1] %>%
+  untar(exdir = "/localssd/imagenet/")
+```
+
+If we are going to be training this model over-and-over using multiple GPUs and even multiple compute instances, we want to make sure we don't waste too much time downloading ImageNet every single time.
+
+The first improvement to consider is getting a faster hard drive. In our case, we locally-mounted an array of SSDs into the `/localssd` path. We then used `/localssd` to extract ImageNet and configured R's temp paths and pins cache to use the SSDs as well. Consult your cloud provider documentation to configure SSDs or take a look at [mlverse/imagenet](https://github.com/mlverse/imagenet).
+
+Next, a well-known approach we can follow is to partition ImageNet into chunks that can be individually downloaded to perform distributed training later on. 
+
+In addition, it is also faster to download ImageNet from a near-by location, ideally from a URL stored within the same datacenter where our cloud instance is located. For this, we can also use pins to register a board with our cloud-provider and then re-upload each partition. Since ImageNet is already partitioned by category, we can easily split ImageNet into multiple zip files and re-upload to your closest datacenter as follows. Make sure the storage bucket is created in the same region as your computing instances.
```

Comment:
```suggestion
In addition, it is also faster to download ImageNet from a nearby location, ideally from a URL stored within the same data center where our cloud instance is located. For this, we can also use pins to register a board with our cloud provider and then re-upload each partition. Since ImageNet is already partitioned by category, we can easily split ImageNet into multiple zip files and re-upload to our closest datacenter as follows. Make sure the storage bucket is created in the same region as your computing instances.
```

Reply:
Is datacenter really one word? The [evolution of a term](https://english.stackexchange.com/questions/68230/life-cycle-life-cycle-or-lifecycle) from open to hyphenated to closed form makes this sort of inherently unanswerable, but I don't think data center is there yet.